airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -152,6 +152,11 @@ def _evaluate_trigger_rule(
             elif tr == TR.ONE_FAILED:
                 if upstream_done and not (failed or upstream_failed):
                     ti.set_state(State.SKIPPED, session)
+            elif tr == TR.NONE_FAILED:
+                if upstream_failed or failed:
+                    ti.set_state(State.UPSTREAM_FAILED, session)
+                elif skipped == upstream:
```

Comment:
@rmn36 @kaxil @ashb @Fokko 

I'm trying to understand the rationale behind this line. E.g. for a dag that looks like this, with just op1 >> op2. When we run the DAG, both op1 and op2 become skipped. The reason is op1 raises `AirflowSkipException`, so op1 is skipped. op2 is set to `none_failed`. But op2 is also skipped because this line checks `skipped == upstream` and marks op2 as skipped.

However, the doc for "Trigger Rule" says:
```none_failed: all parents have not failed (failed or upstream_failed) i.e. all parents have succeeded or been skipped```

So if I interpret this correctly, when all of op2's upstream tasks are skipped, none of them have failed. So op2 should be success rather than skip. 

This may be a feature or a bug. If it's a feature, we need to update the doc. Or if it's a bug we should fix this line.

What do you think?

```
with DAG("skip_exception", schedule_interval=None, start_date=pendulum.parse("20190101")) as dag:
    def raise_skip():
        raise AirflowSkipException
    op1 = PythonOperator(task_id="op1", python_callable=raise_skip)
    op2 = DummyOperator(task_id="op2", trigger_rule="none_failed")
    op1 >> op2
```

Reply:
Pls note that this can be a feature worth preserving, because if someone wants his task op2 to only run if at least some upstream tasks have succeeded (while others can be skipped), he can use "none_failed" to do that because that's exactly what this line does. 

However, if he wants to have his task op2 succeed as long as none of the upstream tasks fail (either skipped for success), he can still use "none_failed" for it. The only workaround he needs to add is to introduce a DummyOperator that always succeeds directly upstream of op2.

So we can keep this feature, but we should make it clear in the doc that if all the upstream tasks are skipped, a task marked with "none_failed" will be skipped. Pls let me know. i don't mind creating a PR to update the doc.

Reply:
> Pls note that this can be a feature worth preserving, because if someone wants his task op2 to only run if at least some upstream tasks have succeeded (while others can be skipped), he can use "none_failed" to do that because that's exactly what this line does.
> 
> However, if he wants to have his task op2 succeed as long as none of the upstream tasks fail (either skipped for success), he can still use "none_failed" for it. The only workaround he needs to add is to introduce a DummyOperator that always succeeds directly upstream of op2.
> 
> So we can keep this feature, but we should make it clear in the doc that if all the upstream tasks are skipped, a task marked with "none_failed" will be skipped. Pls let me know. i don't mind creating a PR to update the doc.

In my opinion, this isn't a reasonable workaround. I'm developing complex DAGs that consist of several tasks with only one upstream dependency-all of which could be potentially skipped. The `none_failed` trigger rule seemed to be the perfect rule for my use case, but anytime I skip a task, skips propagate down until the next task with multiple upstream deps. Wouldn't the reason you mentioned to preserve this feature be satisfied by the ONE_SUCCESS trigger rule? Either way, I believe that `none_failed` communicates "trigger if all upstream are success or skipped" as your docs suggest.

Reply:
cc: @rmn36 @kaxil @ashb @Fokko @yuqian90 

Do you all agree? Should I file a JIRA for this?

Reply:
@ChrisEverling leaving a note here for others who find this conversation, a JIRA ticket was filed for this:  https://issues.apache.org/jira/browse/AIRFLOW-4453